scripts/weekly-summary

```diff
@@ -0,0 +1,374 @@
+            not_fail += 1
+        else:
+            fail += 1
+    return len(tests), not_fail, fail
+
+
+def get_failed(tests):
+    """
+    get_failed returns a dict of failed tests
+
+    Tests may fail multiple times, so they are sorted by their start time
+    and the dict key is the name of the test.
+    """
+    failed = {}
+    for t in tests:
+        if not t["success"]:
+            if t["name"] not in failed:
+                failed[t["name"]] = [t]
+            else:
+                failed[t["name"]].append(t)
+
+    # Sort each failed test list by the start time
+    for t in failed:
+        failed[t] = sorted(failed[t], key=lambda x: x.get("start_time", 0))
+
+    return failed
+
+
+def summary(args, logs):
+    """
+    summary generates a summary of all the tests run in the selected date range
+
+    It returns a string with the summary text
+    """
+    all_data = {}
+    for log in logs:
+        with open(log) as f:
+            data = json.load(f)
+
+            # Group them by group, assume each file is from one group
+            if "group" in data[0]:
+                if data[0]["group"] not in all_data:
+                    all_data[data[0]["group"]] = data
+                else:
+                    all_data[data[0]["group"]].extend(data)
+
+    if args.debug:
+        print(json.dumps(all_data))
+
+    buf = io.StringIO()
+
+    start = args.start.strftime("%m/%d/%Y %H:%M:%S")
+    end = args.end.strftime("%m/%d/%Y %H:%M:%S")
+    print(f"Test Summary Report: {start} -> {end}\n", file=buf)
+
+    # Summary of tests per group
+    for g in sorted(all_data.keys()):
+        total, good, bad = pass_fail(all_data[g])
+        print(f"{g}: Ran {total} tests. {good} passed, {bad} failed.", file=buf)
+    print("\n", file=buf)
+
+    # Test failures per day/group
+    days = {}
+    for g in sorted(all_data.keys()):
+        for t in all_data[g]:
+            if t["success"]:
+                continue
+            day = datetime.fromtimestamp(t.get("start_time", 0)).strftime("%Y%m%d")
+            if day not in days:
+                days[day] = {g: [t["name"]]}
+            else:
+                if g not in days[day]:
+                    days[day][g] = [t["name"]]
+                else:
+                    days[day][g].append(t["name"])
+
+    for day in sorted(days.keys()):
+        if day == "19691231":
+            print("unknown day", file=buf)
+        else:
+            print(datetime.strptime(day, "%Y%m%d").strftime("%m/%d/%Y"), file=buf)
+        for g in sorted(days[day].keys()):
+            print(f"    {g}:", file=buf)
+            for n in sorted(days[day][g]):
+                print(f"        {n}", file=buf)
+        print("\n", file=buf)
+
+    # Gather up detailed failure results
+    for g in sorted(all_data.keys()):
+        print("=" * 40, file=buf)
+        total, good, bad = pass_fail(all_data[g])
+        print(f"{g} summary: Ran {total} tests. {good} passed, {bad} failed.", file=buf)
```

Comment:
I think it would be useful to work with concepts of failure / flake here. I'll go into detail in a standalone comment.

Reply:
Oh my, yes please. Maybe not in this very PR, but eventually we want that pretty badly I think.

That's the main problem of current reporting - retries on failure count two results, which tells me nothing. I want to know "succeeded on retry", not have to parse the logs with my own brain to arrive at that conclusion :-) Also, failures always retry, so for real failures that "correctly re-fail", the final number is artificially inflated (IIRC).

It's one of the reasons why I consider the aggregate numbers after a kickstart tests run meaningless, unless it's a clean success, and instead look at the logs.

Reply:
> It's one of the reasons why I consider the aggregate numbers after a kickstart tests run meaningless, unless it's a clean success, and instead look at the logs.

It should become better when we run it via permian, see the summary added after the aggregate numbers (and also each test finish): https://github.com/rvykydal/kickstart-tests/runs/5497126774?check_suite_focus=true